keystorage/yubihsm2/src/main/java/tech/pegasys/signers/yubihsm/pkcs11/YubiHsmSessionPkcs11.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.signers.yubihsm.pkcs11;
+
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+
+import tech.pegasys.signers.yubihsm.OpaqueDataFormat;
+import tech.pegasys.signers.yubihsm.YubiHsmException;
+import tech.pegasys.signers.yubihsm.YubiHsmSession;
+
+import iaik.pkcs.pkcs11.Module;
+import iaik.pkcs.pkcs11.Session;
+import iaik.pkcs.pkcs11.TokenException;
+import iaik.pkcs.pkcs11.objects.Attribute;
+import iaik.pkcs.pkcs11.objects.ByteArrayAttribute;
+import iaik.pkcs.pkcs11.objects.PKCS11Object;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+
+public class YubiHsmSessionPkcs11 implements YubiHsmSession {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Configuration configuration;
+  private Module module;
+  private Session session;
+
+  public YubiHsmSessionPkcs11(final Configuration configuration) {
+    this.configuration = configuration;
+  }
+
+  public void init() {
+    this.module =
+        Pkcs11ModuleFactory.initPkcs11Module(
+            configuration.getPkcs11ModulePath(), configuration.getPkcs11Conf());
+    this.session = Pkcs11SessionFactory.loginSession(module, configuration.getPin());
+  }
+
+  @Override
+  public Bytes fetchOpaqueData(final short opaqueObjId, final OpaqueDataFormat opaqueDataFormat)
+      throws YubiHsmException {
+    try {
+      initFind(opaqueObjId);
+
+      final byte[] data = findData(opaqueObjId);
+
+      return opaqueDataFormat == OpaqueDataFormat.HEX
```

Comment:
think I'm missing something here - what does it mean if I'm asking for data in hex? Should 'HEX' be 'BINARY' instead?

Reply:
when we store opaque data in yubihsm, we can specify whether we are storing data as HEX (BINARY) or as ASCII (by setting `informat` in yubi-shell command, for instance). 

If we store a BLS key as Hex, it is stored directly as binary. If we store it as ASCII, Yubihsm stores the ascii value of the BLS key i.e. `5e` would be stored as `35 65`. 

Therefore, when we obtain back the data, it can either be stored as BLS binary ... or as Hex String.

Reply:
we can possibly eliminate this step if we mandate that user must always use 'hex' input format to store BLS as Opaque data in YubiHsm for us to work with.